### PR TITLE
Add FP16 and FP64 Supports to PureNcclComunicator

### PR DESCRIPTION
**Pull request by shu65**
This PR adds FP16 and FP64 support to allreduce_grad of PureNcclCommunicator.

### CODE REVIEW

File: chainermn/communicators/_memory_utility.py [hunk 1]
```diff
@@ -70,8 +70,8 @@ def ptr(self):
     def buffer(self, size):
         return self.ffi.buffer(self.ffi.cast('void *', self.memory.ptr), size)
 
-    def array(self, shape, offset=0):
-        return cp.ndarray(shape, memptr=self.memory + offset, dtype=cp.float32)
+    def array(self, shape, offset=0, dtype=np.float32):
```
Comment (kuenishi) [inline chainermn/communicators/_memory_utility.py]: So if we don't want cupy's default behaviour setting dtype as `float64` when `None` is passed, we might better add non-None assertion here.

File: chainermn/communicators/__init__.py [hunk 2]
```diff
@@ -31,6 +34,7 @@ def create_communicator(
           ``hierarchical``, ``two_dimensional``, ``pure_nccl``, or
           ``single_node``)
         mpi_comm: MPI4py communicator
+        allreduce_grad_dtype: Data type of gradient used in All-Reduce
```
[the remark this answers is not in the document]
Comment (kuenishi) [reply chainermn/communicators/__init__.py]: So the behaviour when `None` is passed should be described here, which is to use float type of models.

File: chainermn/communicators/__init__.py [hunk 3]
```diff
@@ -41,6 +45,10 @@ def create_communicator(
         import mpi4py.MPI
         mpi_comm = mpi4py.MPI.COMM_WORLD
 
+    if communicator_name != 'pure_nccl' and allreduce_grad_dtype is not None:
+        raise ValueError(
+            'allreduce_grad_dtype is not supported except for \'pure_nccl\'.')
```
[the remark this answers is not in the document]
Comment (shu65) [reply chainermn/communicators/__init__.py]: Thank you. I will fix this error message according to your comment.

File: chainermn/communicators/__init__.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (shu65) [reply chainermn/communicators/__init__.py]: OK. I will add the description.

File: chainermn/communicators/_memory_utility.py (hunk 1 above)
Comment (shu65) [reply chainermn/communicators/_memory_utility.py]: I I also think so, I will add the assertion.

**Pull request merged**